docs/source/final_attrs.rst

```diff
@@ -0,0 +1,267 @@
+module and class level constants as a way to prevent unintended modification.
+Mypy will prevent further assignments to final names in type-checked code:
+
+.. code-block:: python
+
+   from typing_extensions import Final
+
+   RATE: Final = 3000
+   class Base:
+       DEFAULT_ID: Final = 0
+
+   # 1000 lines later
+
+   RATE = 300  # Error: can't assign to final attribute
+   Base.DEFAULT_ID = 1  # Error: can't override a final attribute
+
+Another use case for final attributes is where a user wants to protect certain
+instance attributes from overriding in a subclass:
+
+.. code-block:: python
+
+   import uuid
+   from typing_extensions import Final
+
+   class Snowflake:
+       """An absolutely unique object in the database"""
+       def __init__(self) -> None:
+           self.id: Final = uuid.uuid4()
+
+   # 1000 lines later
+
+   class User(Snowflake):
+       id = uuid.uuid4()  # Error: can't override a final attribute
+
+Some other use cases might be solved by using ``@property``, but note that
+neither of the above use cases can be solved with it.
+
+.. note::
+
+   This is an experimental feature. Some details might change in later
+   versions of mypy. The final qualifiers are available in the
+   ``typing_extensions`` package available on PyPI.
+
+Syntax variants
+***************
+
+The ``typing_extensions.Final`` qualifier indicates that a given name or
+attribute should never be re-assigned, re-defined, nor overridden. It can be
+used in one of these forms:
+
+
+* You can provide an explicit type using the syntax ``Final[<type>]``. Example:
+
+  .. code-block:: python
+
+     ID: Final[float] = 1
+
+* You can omit the type: ``ID: Final = 1``. Note that unlike for generic
+  classes this is *not* the same as ``Final[Any]``. Here mypy will infer
+  type ``int``.
+
+* In stub files you can omit the right hand side and just write
+  ``ID: Final[float]``.
+
+* Finally, you can define ``self.id: Final = 1`` (also with a type argument),
+  but this is allowed *only* in ``__init__`` methods.
+
+Definition rules
+****************
+
+The are two rules that should be always followed when defining a final name:
+
+* There can be *at most one* final declaration per module or class for
+  a given attribute:
+
+  .. code-block:: python
+
+     from typing_extensions import Final
+
+     ID: Final = 1
+     ID: Final = 2  # Error: "ID" already declared as final
+
+     class SomeCls:
+         id: Final = 1
+         def __init__(self, x: int) -> None:
+             self.id: Final = x  # Error: "id" already declared in class body
+
+  Note that mypy has a single namespace for a class. So there can't be two
+  class-level and instance-level constants with the same name.
+
+* There must be *exactly one* assignment to a final attribute:
+
+  .. code-block:: python
+
+     ID = 1
+     ID: Final = 2  # Error!
+
+     class SomeCls:
+         ID = 1
+         ID: Final = 2  # Error!
```

Comment:
Can you use a different name instead of `ID` for the class variable? Having the name the same distract from the point this example is making regarding duplication. I don't think you need an example showing that a final module global variable and a final class variable don't clash -- it's standard Python semantics (and common knowledge) that those namespaces don't interact.

Reply:
OK, I will update these.